### ISSUE TITLE
Understanding valid mask in the detector loss

### ISSUE THREAD

**Issue by tholmb** [rpautrat/SuperPoint#169]
Thanks again for the great project! I would like to understand the use of valid mask in the detector loss. 

![example2](https://user-images.githubusercontent.com/50668081/93689931-eb426400-fadb-11ea-9854-371a065cf8d9.png)

The upper part of the loss I think that I understand. Sparse_softmax_cross_entropy loss takes logits (B,H/8,W/8,65) from the network just before the softmax layer and labels (B,H/8,W/8) are generated from keypoint map so that if two ground truth corner positions land in the same bin one is selected randomly. The issue is the valid_mask as weights. This is what I think that the code in the red rectangle does when the parameter valid_mask is not None.

![example](https://user-images.githubusercontent.com/50668081/93690110-5fc9d280-fadd-11ea-98c7-648ce8cda437.png)

Homographic augmented data cause bordering artifacts. Valid mask tells the valid area which is first divided with "space to depth" to channels. After that, the AND operation is applied along the channels. In my opinion this leads for most of the augmented samples to zero-weighted matrix and this prevents the learning of neural network. So, are the values of valid_mask always ones or what am I missing? I saw that in the both synthetic shapes dataset and coco dataset pipeline there are first added dummy valid mask and after that the valid mask is computed from the homography matrix H.

Thanks in advance!

**Comment by tholmb** [comment]
Now I got it! I misunderstood the behavior of tf.space_to_depth. Since depth_to_space woks like in https://pic1.zhimg.com/80/v2-c94c787a81c1216d8963f7c173c6f086_1440w.jpg, the code makes sense. Hopefully this is useful for someone else! Sorry to bother you.

**Comment by rpautrat** [comment]
Hi,
Sorry for not answering earlier, but great that you got it by yourself! Indeed, depth_to_space works as in the picture you sent, thanks for posting it with your answer.
